texts/0005-api-refactor.md

```diff
@@ -0,0 +1,79 @@
+
+## Platform Considerations
+
+We can improve the existing desktop APIs, but it is not required to do so. The main goal of this refactor is to actually implement existing functionality on mobile, though improvements are always available and are being discussed for instance on the [capability-based API issue](https://github.com/tauri-apps/tauri/issues/6107).
+
+## Security Considerations
+
+With each API as a separate package, it will be easier to keep the security guarantees from previous audits. Changes will be easier to track and audit.
+
+## Performance Considerations
+
+Performance is not impacted by this change. Existing desktop code runs the same way, and mobile APIs are executed in the native layer so it can leverage the most performant way of doing things.
+
+## Community Considerations
+
+This refactor introduces a massive breaking change. All existing code using the `tauri::api` module or the `@tauri-apps/api` package would break. To mitigate this and simplify the migration path to v2, I propose we move all existing functionality (both the `tauri::api` module and the current Tauri endpoints) to a `legacy` plugin that can be enabled with a Cargo feature. With this plugin, an existing Tauri v1 application could start using the v2 functionalities by just enabling the Cargo feature and changing the `tauri::api` imports to `tauri::legacy`. This solution can also be used for other breaking changes for v2 and beyond, as existing functions that have incompatible new signatures could be moved to a Rust trait that can be imported by the developers to migrate major versions without touching their existing codebase.
+
+# Drawbacks
+
+This change introduces a lot of new packages. Each plugin must have a crate and an NPM package exposing the webview API usage. Keeping both versions in sync is necessary for compatibility, so it introduces complexity for end users. To help users, we are proposing a [`tauri add` CLI command](https://github.com/tauri-apps/tauri/issues/6505).
+
+# Rationale and alternatives
+
+- This design is good for maintainability as it separates application APIs from the core crate. Additionally, it simplifies the mess associated with target detection for the huge list of platforms we must support: macOS, Linux, Windows, Android and iOS.
+- It reduces the current feature flags hell with the allowlist Cargo features. Currently, the tauri crate has over 70 Cargo features related to the allowlist. These flags would be moved to each API plugin crate.
+
+# Prior art
+
+For possible mobile-specific plugins, we can learn a lot from the official Capacitor [plugin collection](https://github.com/ionic-team/capacitor-plugins).
+
+# Unresolved questions
+
+How should we handle versioning? Should we keep minors in sync with the core crate so it is easier to determine incompatibilities? Syncing can be problematic when we have 20+ plugins and we only want to minor bump one of them...
+
+# API list
```

Comment:
This list doesn't include the following JS modules:
- app
- mocks
- tauri
- updater

...and these from the Rust side:
- `api::dir`, `api::file` and `api::path`: are these combined into `file-system`?
- `updater`

Are these candidates to move into the plugin system as well?

Reply:
api::dir and api::file should be removed from the public api imo.

api::path is part of the first path entry i think.

updater is an interesting one, uncoupling it from tauri is nice but we marketed it too much as part of being an integral part of tauri so that i think it should still be exposed as part of the core api 🤔

js-tauri: should be always available (as part of core), especially since it contains `invoke`

js-app: i think this must be part of core because of getTauriVersion, macos app apis could be externalized but may as well keep it there - `app` sounds very core to me anyway

js-mocks: in theory a candidate to externalize but since there is no rust part at all i think it makes sense to keep it in the main package.

Reply:
JS's `app.hide()` and `app.show()` are interesting ones. Because yes, they're technically at the app-level for macOS but then you have things like `window::hide` and `window::show` on the Rust side as well (not sure if also exposed on the JS side).

Reply:
api::dir and api::file should be removed from the public api.

The updater decoupling should be done post v2 IMO, it has too many implications.

js-tauri: core plugin
js-app: core too I think
js-mocks: I believe this one should be moved to a testing plugin. I don't think we should keep using the @tauri-apps/api package anymore, except for legacy stuff and migration paths.

I'll update the list later.